### PR TITLE
Add check for createOptions

### PR DESCRIPTION
**Pull request by SLdragon**

### CODE REVIEW

File: src/iotEdgeExplorer.ts [hunk 1]
```diff
@@ -116,6 +116,35 @@ export class IoTEdgeExplorer extends BaseExplorer {
...
+
+            moduleCreateOptions.forEach( (createOption) => {
+                if (createOption) {
+                    JSON.parse(createOption);
+                }
+            });
+
+            return true;
+        } catch (error) {
+            vscode.window.showErrorMessage(`Cannot parse createOptions of modules: ${error.message}`);
```
Comment (adashen) [inline src/iotEdgeExplorer.ts]: The message should be more clearer to user. And let user know which module has invalid createOptions.

File: src/iotEdgeExplorer.ts [hunk 2]
```diff
@@ -116,6 +116,35 @@ export class IoTEdgeExplorer extends BaseExplorer {
         return true;
     }
 
+    private async isValidCreateOptions(desiredProperties: any): Promise<boolean> {
```
Comment (formulahendry) [inline src/iotEdgeExplorer.ts]: Do we need Promise?

File: src/iotEdgeExplorer.ts [hunk 3]
```diff
@@ -116,6 +116,39 @@ export class IoTEdgeExplorer extends BaseExplorer {
         return true;
     }
 
+    private isValidCreateOptions(desiredProperties: any): boolean {
+        const systemModules = desiredProperties.systemModules;
```
Comment (formulahendry) [inline src/iotEdgeExplorer.ts]: Too many duplicated code, put wrap into a function for the similar code. Or `Cannot convert createOptions of custom module "${moduleName}" as json object: ${error.message}` is enough.

Comment (SLdragon) [reply src/iotEdgeExplorer.ts]: Add a helper function

File: src/iotEdgeExplorer.ts (hunk 2 above)
Comment (SLdragon) [reply src/iotEdgeExplorer.ts]: fixed

File: src/iotEdgeExplorer.ts (hunk 1 above)
Comment (SLdragon) [reply src/iotEdgeExplorer.ts]: Improved

File: src/iotEdgeExplorer.ts [hunk 4]
```diff
@@ -169,7 +199,9 @@ export class IoTEdgeExplorer extends BaseExplorer {
             }
 
             try {
-                const isValid = await this.isValidDeploymentJsonSchema(contentJson);
+                const isValid = await this.isValidDeploymentJsonSchema(contentJson)
+                    && this.isValidCreateOptions(contentJson.modulesContent.$edgeAgent["properties.desired"]);
```
Comment (adashen) [inline src/iotEdgeExplorer.ts]: properties.desired has lots of entries right? Pass the how desired property as modules?

File: src/iotEdgeExplorer.ts [hunk 5]
```diff
@@ -116,6 +116,36 @@ export class IoTEdgeExplorer extends BaseExplorer {
...
+    private isValidCreateOptionsHepler(modules: any): boolean {
+        for (const moduleName in modules) {
+            if (modules.hasOwnProperty(moduleName)) {
+                try {
+                    const createOptions = modules[moduleName].settings.createOptions;
+                    if (createOptions) {
+                        this.checkJsonString(createOptions);
+                    }
+                } catch (error) {
+                    vscode.window.showErrorMessage(`Cannot convert createOptions of module "${moduleName}" as json object: ${error.message}`);
```
Comment (adashen) [inline src/iotEdgeExplorer.ts]: How about change to "createOptions of <module> is not a valid JSON string: ${error.message}"

Comment (SLdragon) [reply src/iotEdgeExplorer.ts]: OK

Comment (SLdragon) [reply src/iotEdgeExplorer.ts]: Updated

**Pull request merged**